File: src/main/java/dev/vankka/simpleast/core/node/StyleNode.java
package dev.vankka.simpleast.core.node;

import lombok.RequiredArgsConstructor;
import lombok.ToString;
import dev.vankka.simpleast.core.TextStyle;

import java.util.List;

/**
 * @param <RC> RenderContext
 * @param <T> Type of Style to apply
 */
@RequiredArgsConstructor
public class StyleNode<RC, T> extends Node<RC> {
    private final List<T> styles;

    public List<T> getStyles() {
        return styles;
    }

    @Override
    public String toString() {
        return "StyleNode{" +
                "styles=" + styles +
                ", children=" + getChildren() + "}";
    }

    public static class Companion {
        public static <RC> StyleNode<RC, TextStyle> createWithText(String content, List<TextStyle> styles) {
            StyleNode<RC, TextStyle> styleNode = new StyleNode<>(styles);
            styleNode.addChild(new TextNode<>(content));
            return styleNode;
        }
    }
}

File: src/main/java/dev/vankka/simpleast/core/node/StyleNode.kt
package dev.vankka.simpleast.core.node

import dev.vankka.simpleast.core.TextStyle

/**
 * @param RC RenderContext
 * @param T Type of Span to apply
 */
open class StyleNode<RC, T>(val styles: List<T>) : Node<RC>() {

  override fun toString() = "${javaClass.simpleName} >\n" +
      getChildren()?.joinToString("\n->", prefix = ">>", postfix = "\n>|") {
        it.toString()
      }

  companion object {

    /**
     * Convenience method for creating a [StyleNode] when we already know what
     * the text content will be.
     */
    @JvmStatic
    fun <RC> createWithText(content: String, styles: List<TextStyle>): StyleNode<RC, TextStyle> {
      val styleNode = StyleNode<RC, TextStyle>(styles)
      styleNode.addChild(TextNode(content))
      return styleNode
    }
  }
}
